src/irr/asset/COBJMeshFileLoader.cpp

```diff
@@ -355,490 +734,377 @@ asset::SAssetBundle COBJMeshFileLoader::loadAsset(io::IReadFile* _file, const as
+    {
+        sprintf(tmpbuf, "layout (set = 3, binding = %u) uniform samplerCube map_refl;\n", j);
+        res += tmpbuf;
+        ++j;
+    }
 
-#ifndef NEW_SHADERS
-	core::smart_refctd_ptr<asset::ICPUTexture> texture;
-	if (texname.size())
-	{
-		auto params = _ctx.inner.params;
-		params.relativeDir = relPath.c_str();
-		auto bundle = interm_getAssetInHierarchy(AssetManager, texname.c_str(), params, _ctx.topHierarchyLevel+ICPUMesh::IMAGEVIEW_HIERARCHYLEVELS_BELOW, _ctx.loaderOverride).getContents();
-        if (bundle.first!=bundle.second) texture = core::smart_refctd_ptr_static_cast<asset::ICPUTexture>(*bundle.first);
-	}
-	if ( texture )
-	{
-		if (type==ETT_COLOR_MAP)
+    constexpr CMTLPipelineMetadata::SMtl::E_MAP_TYPE mandatoryMaps[]
+    {
+        CMTLPipelineMetadata::SMtl::EMP_DIFFUSE,
+        CMTLPipelineMetadata::SMtl::EMP_OPACITY,
+        CMTLPipelineMetadata::SMtl::EMP_SPECULAR,
+        CMTLPipelineMetadata::SMtl::EMP_SHININESS,
+        CMTLPipelineMetadata::SMtl::EMP_AMBIENT,
+        CMTLPipelineMetadata::SMtl::EMP_BUMP
+    };
+    constexpr uint32_t mandatoryMapsSz = sizeof(mandatoryMaps)/sizeof(*mandatoryMaps);
+    //there are a few maps that must be declared in default frag shader (the one with UV) regardless of whetehr they exist or not
+    //in order for it to compile
+    if (_shaderWithUV)
+    {
+        for (uint32_t i = 0u; i < mandatoryMapsSz; ++i)
         {
-			currMaterial->Material.setTexture(0, std::move(texture));
+            const uint32_t mapNum = mandatoryMaps[i];
+            if (!_mtl.maps[mapNum].empty())//really exist, so declaration string was already generated
+                continue;
+
+            sprintf(tmpbuf, "layout (set = 3, binding = %u) uniform sampler2D %s;\n", j, mapNames[mapNum]);
+            res += tmpbuf;
+            ++j;
         }
-		else if (type==ETT_BUMP_MAP)
-		{
-#ifdef _IRR_DEBUG
-            os::Printer::log("Loading OBJ Models with normal maps not supported!\n",ELL_ERROR);
-#endif // _IRR_DEBUG
-			currMaterial->Material.setTexture(1, std::move(texture));
-            //TODO
-			//currMaterial->Material.MaterialType=(video::E_MATERIAL_TYPE)-1;
-			currMaterial->Material.MaterialTypeParam=0.035f;
-		}
-		else if (type==ETT_OPACITY_MAP)
-		{
-			currMaterial->Material.setTexture(3, std::move(texture));
-			//currMaterial->Material.MaterialType=video::EMT_TRANSPARENT_ADD_COLOR;
-		}
-		else if (type==ETT_REFLECTION_MAP)
-		{
-            //TODO
-//						currMaterial->Material.Textures[1] = texture;
-//						currMaterial->Material.MaterialType=video::EMT_REFLECTION_2_LAYER;
-		}
-	}
-#endif
-	return bufPtr;
-}
+    }
 
+    return res;
+}
 
-void COBJMeshFileLoader::readMTL(SContext& _ctx, const char* fileName, const io::path& relPath)
+std::pair<core::smart_refctd_ptr<ICPUSpecializedShader>,core::smart_refctd_ptr<ICPUSpecializedShader>> COBJMeshFileLoader::getShaders(bool _hasUV, const CMTLPipelineMetadata::SMtl& _mtl)
```

Comment:
shouldn't this be in MTL loader?

Reply:
at least the fragment shader thing

Reply:
https://github.com/buildaworldnet/IrrlichtBAW/pull/410#discussion_r367025409

Reply:
views and textures should be handled by MTL loader

Reply:
https://github.com/buildaworldnet/IrrlichtBAW/pull/410#discussion_r367686206

Reply:
still of the opinion textures (descriptor set) and fragment shader should be handled by MTL loader.